qutip/qip/compiler/gatecompiler.py

```diff
@@ -95,23 +126,138 @@ def decompose(self, gates):
+            scheduled_start_time.sort()
+        else:  # no scheduling
+            time_ordered_pos = list(range(0, len(instruction_list)))
+            scheduled_start_time = [0.]
+            for instruction in instruction_list:
+                scheduled_start_time.append(
+                    instruction.duration + scheduled_start_time[-1])
+            scheduled_start_time = scheduled_start_time[:-1]
+
+        # compile
+        # An instruction can be composed from several different pulse elements.
+        # We separate them an assign them to each pulse index.
+        pulse_instructions = [[] for tmp in range(num_pulses)]
+        for instruction, start_time in \
+                zip(instruction_list, scheduled_start_time):
+            for pulse_name, coeff in instruction.pulse_info:
+                if self.pulse_dict is not None:
+                    pulse_ind = self.pulse_dict[pulse_name]
+                else:
+                    pulse_ind = pulse_name
+                pulse_instructions[pulse_ind].append(
+                    (start_time, instruction.tlist, coeff))
+
+        compiled_tlist = [[[0.]] for tmp in range(num_pulses)]
+        compiled_coeffs = [[] for tmp in range(num_pulses)]
+        for pulse_ind in range(num_pulses):
+            for start_time, tlist, coeff in pulse_instructions[pulse_ind]:
+                if np.isscalar(tlist):
+                    # a single constant rectanglar pulse, where
+                    # tlist and coeff are just float numbers
+                    step_size = tlist
+                    temp_tlist = np.array([tlist])
+                    coeff = np.array([coeff])
+                elif len(tlist) - 1 == len(coeff):
+                    # discrete pulse
+                    step_size = tlist[1] - tlist[0]
+                    temp_tlist = np.asarray(tlist)[1:]
+                    coeff = np.asarray(coeff)
+                elif len(tlist) == len(coeff):
+                    # continuos pulse
+                    step_size = tlist[1] - tlist[0]
+                    if compiled_coeffs[pulse_ind]:  # not first pulse
+                        temp_tlist = np.asarray(tlist)[1:]
+                        coeff = np.asarray(coeff)[1:]
+                    else:  # first pulse
+                        temp_tlist = np.asarray(tlist)[1:]
+                        coeff = np.asarray(coeff)
+                else:
+                    raise ValueError(
+                        "The shape of the compiled pulse is not correct.")
+                # If there is a idling time between the last pulse and
+                # the current one, we need to add zeros in between.
+                # We add sufficient number of zeros at the begining
+                # and the end of the idling to prevent wrong cubic spline.
+                if np.abs(start_time - compiled_tlist[pulse_ind][-1][-1]) \
+                        > step_size * 1.0e-6:
```

Comment:
When initializing `compiled_tlist` I made sure that it is never empty: `
compiled_tlist = [[[0.]] for tmp in range(num_controls)]
`

The 3-level nested list is for 
- control_pulse_index
- instruction
- tlist for each instruction


Here I need to know where the last time step ends. And I don't want to check every time if this is empty. So I added a 0 there and if it is no instruction for that control pulse, I'll get a 0.

Reply:
It's not super obvious that the initial value you assign to the output variable `compiled_tlist` should be important, since the rest of the loop is about overwriting all of that, and three-level nested loops are difficult enough at the best of times.

For a maintainer looking at the loop in six months' time it seems like a classic off-by-one access error on the first step, especially as the initialisation is about 30--40 lines higher up.  It'd be clearer if the last time step was stored in a variable that's just called `last_step_end` or something which is initialised to `0` before the loop, not least because then you can read words rather than array accesses and you won't need a special sentinel value in the initialisation.

I get that it works now, but it just seems kind of fragile and difficult to read the next time someone needs to modify the code.

Reply:
> It'd be clearer if the last time step was stored in a variable that's just called last_step_end or something which is initialised to 0 before the loop

This sounds prominent. I'll work on this.